File: LeetCodeHot100/_617_合并二叉树/mergeTrees.py
'''
给你两棵二叉树： root1 和 root2 。

想象一下，当你将其中一棵覆盖到另一棵之上时，两棵树上的一些节点将会重叠（而另一些不会）。你需要将这两棵树合并成一棵新二叉树。合并的规则是：如果两个节点重叠，那么将这两个节点的值相加作为合并后节点的新值；否则，不为 null 的节点将直接作为新二叉树的节点。

返回合并后的二叉树。

注意: 合并过程必须从两个树的根节点开始。

示例 1：
输入：root1 = [1,3,2,5], root2 = [2,1,3,null,4,null,7]
输出：[3,4,5,5,4,null,7]

示例 2：
输入：root1 = [1], root2 = [1,2]
输出：[2,2]
 

提示：

两棵树中的节点数目在范围 [0, 2000] 内
-104 <= Node.val <= 104
'''


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def mergeTrees(self, root1: TreeNode, root2: TreeNode) -> TreeNode:
        # 方法一：使用递归

        # 如果root1和root2其中一个为None，则返回另外一个节点
        if root1 == None:
            return root2
        elif root2 == None:
            return root1
        
        # root1和root2都不为None
        else:
            # 创建新节点
            newNode = TreeNode(root1.val + root2.val)
            newNode.left = self.mergeTrees(root1.left, root2.left)
            newNode.right = self.mergeTrees(root1.right, root2.right)
            return newNode

        
        # # 方法二：
        # # 使用广度优先遍历
        # # 将root1没有，root2有的子树，链接到root1，并且root1和root2都不需要遍历这一部分
        # # 将root1和root2都存在的子树的每个节点的数据进行相加，同时遍历这一部分
        # if root1 == None:
        #     return root2
        # elif root2 == None:
        #     return root1
        # else:
        #     queue1 = []         # root1的队列
        #     queue2 = []         # root2的队列
        #     queue1.append(root1)
        #     queue2.append(root2)
        #     while queue1 != [] and queue2 != []:
        #         cur1 = queue1.pop(0)
        #         cur2 = queue2.pop(0)

        #         # root1和root2都有的节点，每个节点的数据进行相加，同时遍历他们的左右子树
        #         cur1.val += cur2.val

        #         # 左孩子
        #         # 如果root1和root2的左孩子都有，则遍历这一部分
        #         if cur1.left != None and cur2.left != None:
        #             queue1.append(cur1.left)
        #             queue2.append(cur2.left)
                
        #         # 如果root1的左孩子没有，root2的左孩子有，则将root2的左孩子复制添加到root1中，不遍历这一部分，因为他们的值不需要再次相加
        #         elif cur1.left == None and cur2.left != None:
        #             cur1.left = cur2.left 
                
        #         # 如果root1的左孩子有， root2的左孩子没有，则root1和root2的左孩子都不需要遍历，因为root1的这一部分不需要相加，同时因为返回的是root1，所以也不需要复制到root2中
        #         # 如果root1和root2的左孩子都没有，则不需要遍历他们的左孩子
        #         else:
        #             pass

        #         # 右孩子
        #         # 同理左孩子
        #         if cur1.right != None and cur2.right != None:
        #             queue1.append(cur1.right)
        #             queue2.append(cur2.right)
        #         elif cur1.right == None and cur2.right != None:
        #             cur1.right = cur2.right
        #         else:
        #             pass

        # # 返回root1
        # return root1
            

File: LeetCodeHot100/_617_合并二叉树/mergeTrees.c
#include<stdio.h>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


struct TreeNode* mergeTrees(struct TreeNode* root1, struct TreeNode* root2){
    // 方法1：使用递归：
    // 1. 如果root1没有孩子，则返回root2
    if (root1 == NULL) 
    {
        return root2;
    }

    // 2. 如果root2没有孩子，则返回root1
    else if (root2 == NULL) 
    {
        return root1;
    }
    
    // 3. 如果root1和root2都有孩子
    else 
    {
        // 3.1 分配一个大小为TreeNode的内存
        struct TreeNode *merge = malloc(sizeof(struct TreeNode));
        merge->val = root1->val + root2->val;
        // 3.2 递归，得到左孩子和右孩子
        merge->left = mergeTrees(root1->left, root2->left);
        merge->right = mergeTrees(root1->right, root2->right);
        return merge;
    }

}

struct TreeNode* mergeTrees2(struct TreeNode* root1, struct TreeNode* root2) {
    // 方法二：使用队列、层序遍历
    // 1. 如果root1为NULL，则返回root2
    if (root1 == NULL) {
        return root2;
    }
    // 2. 如果root2为NULL，则返回root1
    else if (root2 == NULL){
        return root1;
    }
    // 3. 如果root1和root2都不为NULL
    else {
        struct TreeNode *merge = malloc(sizeof(struct TreeNode));
        merge->val = root1->val + root2->val;

        // 定义队列
        struct TreeNode **queue1 = malloc(sizeof(struct TreeNode *) * 2001);
        struct TreeNode **queue2 = malloc(sizeof(struct TreeNode *) * 2001);
        struct TreeNode **q = malloc(sizeof(struct TreeNode *) * 4002);

        // 定义变量
        int cur_deep = 0;       // 当前的层数
        int max_deep = 0;       // 能够遍历的最大层数，即遍历的下一层。只有当root1和root2的孩子都同时存在时，才能够增加max_deep的值，因为其他情况都不需要再遍历了。当cur_deep小于max_deep时才符合要求。 

        // 初始化队列
        queue1[max_deep] = root1;
        queue2[max_deep] = root2;
        q[max_deep] = merge;
        max_deep++;

        while (cur_deep < max_deep){

            // 当前层数的当前节点
            struct TreeNode *cur_1 = queue1[cur_deep];     
            struct TreeNode *cur_2 = queue2[cur_deep];
            struct TreeNode* node = q[cur_deep];        // merge的树中当前节点

            cur_deep++;            // 层数加一

            // 3.1 对于左孩子
            // (1). 如果root1和root2的左孩子都存在，则他们的值相加
            if (cur_1->left != NULL && cur_2->left != NULL) {
                // 申请一个新的地址存放node的左孩子节点
                struct TreeNode* leftNode = malloc(sizeof(struct TreeNode));
                leftNode->val = cur_1->left->val + cur_2->left->val;  // root1和root2的值相加
                node->left = leftNode;                    
                // 将遍历到的新的节点加入到队列中
                queue1[max_deep] = cur_1->left;
                queue2[max_deep] = cur_2->left;
                q[max_deep] = node->left;
                max_deep++;                 // 最大深度加1
            }
            // (2). 如果root1的左孩子存在，root2的左孩子不存在
            if (cur_1->left != NULL && cur_2->left == NULL) {
                node->left = cur_1->left;
            }
            // (3). 如果root1的左孩子不存在，root2的左孩子存在
            if (cur_1->left == NULL && cur_2->left != NULL) {
                node->left = cur_2->left;
            }
            // (4). 如果root1和root2的左孩子都不存在，则为NULL
            if (cur_1->left == NULL && cur_2->left == NULL) {
                node->left = NULL;
            }

            // 右孩子同理
            // (1). 如果root1和root2的右孩子都存在
            if (cur_1->right != NULL && cur_2->right != NULL) {
                // 生成一个新的节点存放merge的右孩子
                struct TreeNode* rightNode = malloc(sizeof(struct TreeNode));
                // 新的节点的值
                rightNode->val = cur_1->right->val + cur_2->right->val;
                // 将新的节点rightNode加入到merge树当前节点node的右孩子中
                node->right = rightNode;
                // 将还需要遍历的节点加入到队列中
                queue1[max_deep] = cur_1->right;
                queue2[max_deep] = cur_2->right;
                q[max_deep] = rightNode;
                max_deep++;
            }
            // (2). 如果root1的右孩子存在，root2的右孩子不存在
            if (cur_1->right != NULL && cur_2->right == NULL) {
                node->right = cur_1->right;
            }
            // (3). 如果root1的右孩子不存在，root2的右孩子存在
            if (cur_1->right == NULL && cur_2->right != NULL) {
                node->right = cur_2->right;
            }
            // (4). 如果root1和root2的右孩子都不存在
            if (cur_1->right == NULL && cur_2->right == NULL) {
                node->right = NULL;
            }
        }

        return merge;
    }

}
